core/src/coproduct.rs

```diff
@@ -405,6 +441,112 @@ where
     }
 }
 
+/// Trait for extracting a subset of the possible types in a coproduct.
+///
+/// This is basically [`uninject`] on steroids.  It lets you remove a number
+/// of types from a coproduct at once, leaving behind the remainder in an `Err`.
+/// For instance, one can extract `Coprod!(C, A)` from `Coprod!(A, B, C, D)`
+/// to produce `Result<Coprod!(C, A), Coprod!(B, D)>`.
+///
+/// Each type in the extracted subset is required to be part of the input coproduct.
+///
+/// [`uninject`]: ../traits.CoprodUninjector.html
+///
+/// # Example
+///
+/// Basic usage:
+///
+/// ```
+/// # #[macro_use] extern crate frunk_core;
+/// # use frunk_core::coproduct::*;
+/// # fn main() {
+/// type I32BoolF32 = Coprod!(i32, bool, f32);
+///
+/// let co: Result<Coprod!(i32, f32), _> = I32BoolF32::inject(42_f32).subset();
+/// assert!(co.is_ok());
+///
+/// let co: Result<Coprod!(i32, f32), _> = I32BoolF32::inject(true).subset();
+/// assert!(co.is_err());
+/// # }
+/// ```
+///
+/// Like `uninject`, `subset` can be used for exhaustive matching,
+/// with the advantage that it can remove more than one type at a time:
+///
+/// ```
+/// # #[macro_use] extern crate frunk_core;
+/// # use frunk_core::coproduct::*;
+/// # fn main() {
+/// fn handle_stringly_things(co: Coprod!(&'static str, String)) -> String {
+///     co.fold(hlist![
+///         |s| format!("&str {}", s),
+///         |s| format!("String {}", s),
+///     ])
+/// }
+///
+/// fn handle_countly_things(co: Coprod!(u32)) -> String {
+///     co.fold(hlist![
+///         |n| vec!["."; n as usize].concat(),
+///     ])
+/// }
+///
+/// type Anything = Coprod!(String, u32, &'static str);
+/// fn handle_anything(co: Anything) -> String {
+///
+///     // co is currently Coprod!(String, u32, &'static str)
+///     let co = match co.subset().map(handle_stringly_things) {
+///         Ok(s) => return s,
+///         Err(co) => co,
+///     };
+///
+///     // now co is Coprod!(u32)
+///     let co = match co.subset().map(handle_countly_things) {
+///         Ok(s) => return s,
+///         Err(co) => co,
+///     };
+///
+///     // now co is empty;
+///     match co { /* unreachable */ }
+/// }
+///
+/// assert_eq!(handle_anything(Anything::inject("hello")), "&str hello");
+/// assert_eq!(handle_anything(Anything::inject(String::from("World!"))), "String World!");
+/// assert_eq!(handle_anything(Anything::inject(4)), "....");
+/// # }
+/// ```
+pub trait CoproductSubsetter<Targets, Indices>: Sized {
```

Comment:
I wonder if this needs a separate trait or if it would be possible to simply do this as impls of `CoproductUninjector` with a label type?

Reply:
What do you mean by label type? In general I am fearful of having these be the same method (especially if it is entirely driven by type inference) because of the ambiguities created when you have a coproduct containing coproducts.

Reply:
Take a look at:
+ https://docs.rs/frunk_core/0.0.23/frunk_core/hlist/trait.LiftInto.html
+ https://docs.rs/frunk_core/0.0.23/frunk_core/hlist/trait.LiftFrom.html
+ https://docs.rs/frunk_core/0.0.23/frunk_core/hlist/struct.Suffixed.html -- here's the index type.

Inference problems is a good point 👍

Reply:
@Centril That is the most criminally underdocumented and bizarre feature I have ever seen.

(Well, second most bizarre.  The *most* bizarre feature I've seen are the impls of `From` and `Into` for tuples of the wrong size, sitting there right next to the impls for tuples of matching size)

* The documentation of `lift_from` shows HLists being produced from single values.
* Why on earth is this same function also used to generate an HList from a prefix?  Why a prefix and not a suffix?  Why not an arbitrarily ordered subset?
* It shares my aforementioned concern about ambiguous input types.

Reply:
> That is the most criminally underdocumented and bizarre feature I have ever seen.

Feel free to improve the documentation. The docs for `LiftInto` shows Hlists being produced from multiple values, but its out of sync with `LiftFrom`.

> Why a prefix and not a suffix? Why not an arbitrarily ordered subset?

Prefixes are supported afaik.
Arbitrarily ordered subsets never crossed my mind; If you can improve on that, it would be nice.